swabbie-redis/src/main/kotlin/com/netflix/spinnaker/swabbie/redis/RedisResourceUseTrackingRepository.kt

```diff
@@ -0,0 +1,113 @@
+import java.time.Clock
+import java.time.Duration
+
+@Component
+class RedisResourceUseTrackingRepository(
+  @Qualifier("redisClientSelector") redisClientSelector: RedisClientSelector,
+  private val objectMapper: ObjectMapper,
+  private val clock: Clock,
+  swabbieProperties: SwabbieProperties
+) : ResourceUseTrackingRepository {
+
+  private val log = LoggerFactory.getLogger(javaClass)
+  private val LAST_SEEN = "{swabbie:lastseen}"
+  private val READY_FOR_CLEANING = "{swabbie:readyforcleaning}"
+  private val INIT_TIME = "{swabbie:inittime}"
+  private val outOfUseThresholdDays = swabbieProperties.outOfUseThresholdDays
+
+  private val redisClientDelegate: RedisClientDelegate = redisClientSelector.primary("default")
+
+  init {
+    log.info("Using ${javaClass.simpleName}")
+
+    redisClientDelegate.withCommandsClient { client ->
+      client.setnx(INIT_TIME, clock.instant().toEpochMilli().toString())
+    }
+  }
+
+  override fun hasCompleteData(): Boolean {
+    return redisClientDelegate.run {
+      this.withCommandsClient<String> { client ->
+        client.get(INIT_TIME)
+      }
+    }.toLong() < clock.instant().minus(Duration.ofDays(outOfUseThresholdDays.toLong())).toEpochMilli()
+  }
+
+  override fun recordUse(resourceIdentifier: String, usedByResourceIdentifier: String) {
+    redisClientDelegate.withCommandsClient { client ->
+      client.hset(
+        LAST_SEEN,
+        resourceIdentifier,
+        objectMapper.writeValueAsString(
+          LastSeenInfo(resourceIdentifier, usedByResourceIdentifier, clock.instant().toEpochMilli()))
+      )
+      // add to sorted set by score
+      // score = day when resource will be ready to delete
+      // if resource is seen in next X days, resource will be updated
+      client.zadd(READY_FOR_CLEANING, plusXdays(outOfUseThresholdDays).toDouble(), resourceIdentifier)
+    }
+  }
+
+  override fun getUnused(): List<LastSeenInfo> {
+    val keys = redisClientDelegate.run {
+      this.withCommandsClient<Set<String>> { client ->
+        client.zrangeByScore(READY_FOR_CLEANING, 0.0, clock.instant().toEpochMilli().toDouble())
+      }
+    }
+    return hydrateLastSeen(keys)
+  }
+
+  private fun hydrateLastSeen(keys: Set<String>): List<LastSeenInfo> {
+    if (keys.isEmpty()) return emptyList()
+    return redisClientDelegate.run {
+      this.withCommandsClient<Set<String>> { client ->
+        client.hmget(LAST_SEEN, *keys.toTypedArray()).toSet()
+      }.map { json ->
+        objectMapper.readValue<LastSeenInfo>(json)
+      }
+    }
+  }
+
+  override fun isUnused(resourceIdentifier: String): Boolean {
+    return redisClientDelegate.run {
+      this.withCommandsClient<Double> { client ->
+        client.zscore(READY_FOR_CLEANING, resourceIdentifier)
+      }
+    }.toLong() < clock.instant().toEpochMilli()
+  }
+
+  fun plusXdays(days: Int): Long {
+    val secondsToAdd = (60 * 60 * 24 * days).toLong()
+    return clock.instant().plusSeconds(secondsToAdd).toEpochMilli()
```

Comment:
If you're doing maths to calculate time, there's probably a better way:

```kotlin
clock.instant().plus(days, ChronoUnit.DAYS).toEpochMilli()
```

Reply:
Fixing and thank you!